lms/lib/comment_client/utils.py

```diff
@@ -192,3 +196,60 @@ def check_forum_heartbeat():
             return 'forum', False, res.get('check', 'Forum heartbeat failed')
     except Exception as fail:
         return 'forum', False, unicode(fail)
+
+
+# User not found sentinel. Keep it in the cache to avoid hitting the database
+# on non-existent or not found user.
+user_not_found = object()
+
+
+def annotate_with_first_name(obj, user_id_field="user_id"):
+    """Annotate an object with the author first name in place."""
+    annotate_dict_with_first_name(obj.attributes, user_id_field=user_id_field)
+
+
+def annotate_dict_with_first_name(attributes, user_id_field="user_id"):
+    """Annotate dict and all its children with user_first_name."""
+    first_name = attributes.get("user_first_name")
+    if not first_name:
+        first_name = get_first_name(attributes.get(user_id_field))
+        attributes["user_first_name"] = first_name
+
+    child_keys = ["children", "endorsed_responses", "non_endorsed_responses"]
+
+    for child_key in child_keys:
+        children = attributes.get(child_key)
+        if not isinstance(children, list):
+            continue
+        for child in children:
+            if not isinstance(child, dict):
+                continue
+            # All children objects have a "user_id" field, so we hardcode the value
+            annotate_dict_with_first_name(child, user_id_field="user_id")
+
+
+@request_cached
+def get_first_name(user_id):
+    """Return user's first name by its ID.
+
+    The function uses the Django cache functionality to avoid hitting the database
+    too often. The name is cached for 1 hour (3600 seconds), expires, but never
+    gets invalidated.
+    """
+    if not user_id:
+        return None
+
+    cache_timeout = 3600
+    cache_key = 'comment_client.get_first_name.{}'.format(user_id)
+    first_name = cache.get(cache_key)
+    if first_name is user_not_found:
+        return None
+
+    if first_name is None:
+        try:
+            first_name = User.objects.get(id=user_id).profile.first_name
+        except User.NotFound:
```

Comment:
This is causing Jenkins errors 
`[](http://build.edraak.cc/blue/organizations/jenkins/edraak-platform/detail/PR-263/6/pipeline/#step-256-log-1427)[](http://build.edraak.cc/blue/organizations/jenkins/edraak-platform/detail/PR-263/6/pipeline/#step-256-log-1428)File "/home/jenkins/workspace/edraak-platform_PR-263/lms/lib/comment_client/utils.py", line 251, in get_first_name

[2022-03-18T18:30:16.251Z]     except User.NotFound:

[2022-03-18T18:30:16.251Z] AttributeError: type object 'User' has no attribute 'NotFound'`

Reply:
Oh, thanks for the catch 😬